### PR TITLE
Add check for adding param to method

### LINKED ISSUE #750: Why is this not a breaking change?
```diff
 class HttpClient
 {
-    public function request(string $method, string $url)
+    public function request(string $method, string $url, array $options = [])
     {
 
     }
```

My thoughts are: If I extended the `HttpClient` before this change, then after the change I would get this error:

```
Fatal error: Declaration of ConcreteHttpClient::request(string $method, string $url) must be compatible with HttpClient::request(string $method, string $url, array $options = [])
```

### PR DESCRIPTION
**Pull request by AydinHassan**
Heyy,

We recently encountered #750 and I had a quick stab at implementing it, to start the conversation off. I guess we should also support removing. Feedback welcome!

Fixes #750

### CODE REVIEW

File: test/unit/DetectChanges/BCBreak/MethodBased/MethodParameterAddedTest.php [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+            ,
+            $astLocator,
+        );
+
+        $toLocator = new StringSourceLocator(
+            <<<'PHP'
+<?php
+
+class TheClass {
+    public function addedParameter(array $options = []) {}
```
Comment (Ocramius) [inline test/unit/DetectChanges/BCBreak/MethodBased/MethodParameterAddedTest.php:124]: This is a breakage because child classes will have to adapt, right?

Comment (AydinHassan) [reply test/unit/DetectChanges/BCBreak/MethodBased/MethodParameterAddedTest.php:124]: exactly!

File: src/DetectChanges/BCBreak/MethodBased/MethodParameterAdded.php [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+use Roave\BackwardCompatibility\Change;
+use Roave\BackwardCompatibility\Changes;
+use Roave\BetterReflection\Reflection\ReflectionMethod;
+use Roave\BetterReflection\Reflection\ReflectionParameter;
+
+use function array_diff;
+use function array_map;
+
+/**
+ * Adding a parameter to a method on a non-final class is a BC break
```
Comment (Ocramius) [inline src/DetectChanges/BCBreak/MethodBased/MethodParameterAdded.php]: Let's add the "why" in this comment, and then we can merge :-)

Comment (AydinHassan) [reply src/DetectChanges/BCBreak/MethodBased/MethodParameterAdded.php]: Nice! Updated :)

**Review by Ocramius — APPROVED** [review]
:ship: thanks @AydinHassan!

**Comment by Ocramius** [comment]
@AydinHassan uh-oh, some more tests to add to cover the last introduced mutations

```diff
Warning: Escaped Mutant for Mutator "LogicalOr":

--- Original
+++ New
@@ @@
 {
     public function __invoke(ReflectionMethod $fromMethod, ReflectionMethod $toMethod) : Changes
     {
-        if ($fromMethod->getDeclaringClass()->isFinal() || $fromMethod->isPrivate()) {
+        if ($fromMethod->getDeclaringClass()->isFinal() && $fromMethod->isPrivate()) {
             return Changes::empty();
         }
         $added = array_diff(array_map(static fn(ReflectionParameter $param) => $param->getName(), $toMethod->getParameters()), array_map(static fn(ReflectionParameter $param) => $param->getName(), $fromMethod->getParameters()));


Warning: Escaped Mutant for Mutator "UnwrapArrayDiff":

--- Original
+++ New
@@ @@
         if ($fromMethod->getDeclaringClass()->isFinal() || $fromMethod->isPrivate()) {
             return Changes::empty();
         }
-        $added = array_diff(array_map(static fn(ReflectionParameter $param) => $param->getName(), $toMethod->getParameters()), array_map(static fn(ReflectionParameter $param) => $param->getName(), $fromMethod->getParameters()));
+        $added = array_map(static fn(ReflectionParameter $param) => $param->getName(), $toMethod->getParameters());
         if ($toMethod->getNumberOfParameters() > $fromMethod->getNumberOfParameters()) {
             return Changes::fromList(...array_map(static fn(string $paramName): Change => Change::added(Str\format('Parameter %s was added to Method %s() of class %s', $paramName, $fromMethod->getName(), $fromMethod->getDeclaringClass()->getName())), $added));
         }


Warning: Escaped Mutant for Mutator "UnwrapArrayMap":

--- Original
+++ New
@@ @@
         if ($fromMethod->getDeclaringClass()->isFinal() || $fromMethod->isPrivate()) {
             return Changes::empty();
         }
-        $added = array_diff(array_map(static fn(ReflectionParameter $param) => $param->getName(), $toMethod->getParameters()), array_map(static fn(ReflectionParameter $param) => $param->getName(), $fromMethod->getParameters()));
+        $added = array_diff(array_map(static fn(ReflectionParameter $param) => $param->getName(), $toMethod->getParameters()), $fromMethod->getParameters());
         if ($toMethod->getNumberOfParameters() > $fromMethod->getNumberOfParameters()) {
             return Changes::fromList(...array_map(static fn(string $paramName): Change => Change::added(Str\format('Parameter %s was added to Method %s() of class %s', $paramName, $fromMethod->getName(), $fromMethod->getDeclaringClass()->getName())), $added));
         }


Warning: Escaped Mutant for Mutator "GreaterThan":

--- Original
+++ New
@@ @@
             return Changes::empty();
         }
         $added = array_diff(array_map(static fn(ReflectionParameter $param) => $param->getName(), $toMethod->getParameters()), array_map(static fn(ReflectionParameter $param) => $param->getName(), $fromMethod->getParameters()));
-        if ($toMethod->getNumberOfParameters() > $fromMethod->getNumberOfParameters()) {
+        if ($toMethod->getNumberOfParameters() >= $fromMethod->getNumberOfParameters()) {
             return Changes::fromList(...array_map(static fn(string $paramName): Change => Change::added(Str\format('Parameter %s was added to Method %s() of class %s', $paramName, $fromMethod->getName(), $fromMethod->getDeclaringClass()->getName())), $added));
         }
         return Changes::empty();
     }
 }

Warning:  Dashboard report has not been sent: The current process is a pull request build
```

This can be tested by expanding your data provider:

 * add examples that remove parameters -> shouldn't trip this comparator
 * add examples with `private` methods -> shouldn't trip this comparator

**Comment by AydinHassan** [comment]
@Ocramius I tried adding a few cases which I've just pushed. I also added some other tests that I since deleted but it doesn't seem to fix the escaped mutants. Sorry I don't have any experience with Infection and looks like I am fundamentally missing something. Whatever I change seems to have no effect :D 

edit: i'm a dumbass missed adding the provider cases

edit 2: Ok I managed to kill a few but I'm a bit lost with the `GreaterThan` mutant

edit 3: Fixed the last one by deleting code 🔪 

@Ocramius could you so kindly approve the workflow :)

**Comment by Ocramius** [comment]
> @Ocramius I tried adding a few cases which I've just pushed. I also added some other tests that I since deleted but it doesn't seem to fix the escaped mutants. Sorry I don't have any experience with Infection and looks like I am fundamentally missing something. Whatever I change seems to have no effect :D
> 
> edit: i'm a dumbass missed adding the provider cases
> 
> edit 2: Ok I managed to kill a few but I'm a bit lost with the `GreaterThan` mutant
> 
> edit 3: Fixed the last one by deleting code 🔪
> 
> @Ocramius could you so kindly approve the workflow :)

and now you know how it works :-)

**Review by Ocramius — APPROVED** [review]
:ship: excellent!

**Comment by AydinHassan** [comment]
thanks for the guidance  and speed @Ocramius ! 🏎️

**Pull request merged**